File: mobile/ios/FUN/FUN/ViewModels/FeedViewModel.swift
/**
 * FUN App - Feed ViewModel
 * Manages feed content and playback
 */

import Foundation
import Combine

@MainActor
class FeedViewModel: ObservableObject {
    @Published var episodes: [FeedEpisode] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var hasMore = true
    
    private var currentPage = 1
    private let pageLimit = 10
    private let apiClient = APIClient.shared
    private var episodesWatchedCount = 0
    private let interstitialFrequency = 3  // Show after every 3 episodes
    
    // MARK: - Data Loading
    
    func loadFeed(refresh: Bool = false) async {
        if refresh {
            currentPage = 1
            episodes = []
            hasMore = true
        }
        
        guard hasMore && !isLoading else { return }
        
        isLoading = true
        errorMessage = nil
        
        do {
            let response: FeedResponse = try await apiClient.request(
                .feed(page: currentPage, limit: pageLimit)
            )
            
            if refresh {
                episodes = response.episodes
            } else {
                episodes.append(contentsOf: response.episodes)
            }
            
            hasMore = currentPage < response.pagination.pages
            currentPage += 1
            isLoading = false
        } catch {
            errorMessage = error.localizedDescription
            isLoading = false
        }
    }
    
    func loadMore() async {
        await loadFeed()
    }
    
    func refresh() async {
        await loadFeed(refresh: true)
    }
    
    // MARK: - Interactions
    
    func toggleLike(seriesId: String) async {
        do {
            try await apiClient.requestWithoutResponse(
                .likeSeries(id: seriesId),
                method: .post
            )
            
            // Update local state
            if let index = episodes.firstIndex(where: { $0.series.id == seriesId }) {
                // Toggle like state locally
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }
    
    // MARK: - Interstitial Ads
    
    func onEpisodeCompleted() {
        episodesWatchedCount += 1
        
        // Show interstitial after every 3 episodes
        if episodesWatchedCount % interstitialFrequency == 0 {
            AdManager.shared.showInterstitialAd()
        }
    }
    
    func onEpisodeStarted() {
        // Track episode view
    }
}

File: mobile/android/FUN/app/src/main/java/com/fun/app/ui/viewmodels/FeedViewModel.kt
/**
 * FUN App - Feed ViewModel (Android)
 */

package com.fun.app.ui.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.fun.app.FunApplication
import com.fun.app.core.network.ApiClient
import com.fun.app.core.network.NetworkResult
import com.fun.app.core.network.safeApiCall
import com.fun.app.data.models.FeedEpisode
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class FeedViewModel(
    private val apiClient: ApiClient
) : ViewModel() {
    
    private val _episodes = MutableStateFlow<List<FeedEpisode>>(emptyList())
    val episodes: StateFlow<List<FeedEpisode>> = _episodes.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    private var currentPage = 1
    private val pageLimit = 10
    private var hasMore = true
    private var episodesWatchedCount = 0
    private val interstitialFrequency = 3  // Show after every 3 episodes
    
    fun loadFeed(refresh: Boolean = false) {
        if (refresh) {
            currentPage = 1
            _episodes.value = emptyList()
            hasMore = true
        }
        
        if (!hasMore && !refresh) return
        
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            when (val result = safeApiCall {
                apiClient.apiService.getFeed(page = currentPage, limit = pageLimit)
            }) {
                is NetworkResult.Success -> {
                    if (refresh) {
                        _episodes.value = result.data.episodes
                    } else {
                        _episodes.value = _episodes.value + result.data.episodes
                    }
                    
                    hasMore = currentPage < result.data.pagination.pages
                    currentPage++
                }
                is NetworkResult.Error -> {
                    _errorMessage.value = result.message
                }
                else -> {}
            }
            
            _isLoading.value = false
        }
    }
    
    fun loadMore() {
        loadFeed(refresh = false)
    }
    
    fun refresh() {
        loadFeed(refresh = true)
    }
    
    fun onEpisodeCompleted() {
        episodesWatchedCount++
        
        // Show interstitial after every 3 episodes
        if (episodesWatchedCount % interstitialFrequency == 0) {
            // Note: This requires Activity context, will be triggered from UI layer
        }
    }
    
    fun onEpisodeStarted() {
        // Track episode view
    }
    
    companion object {
        val Factory: ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                return FeedViewModel(
                    apiClient = FunApplication.instance.apiClient
                ) as T
            }
        }
    }
}
